src/analysis/souffle/tests/test_taint_with_auth_logic.dl

```diff
@@ -0,0 +1,101 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#include "../taint_with_auth_logic.dl"
+#include "../flat_graph_ir.dl"
+#include "../edges_from_graph.dl"
+
+// This is a datalog file that is meant to test taint_with_auth_logic.dl. It 
+// should eventually be replaced by or otherwise used as a part of an
+// automated test.
+
+//-----------------------------------------------------------------------------
+// Program Manifests
+//-----------------------------------------------------------------------------
+
+// Module 1
+//      reads: 
+//          some_struct{ a, b }
+//          other_input
+//      writes:
+//          mod1out
+bindFlatParticle("module1", $FlatParticle(
+    /* reads */
+    $Cons("some_struct.a", $Cons("some_struct.b",
+        $Cons("other_input", $Nil()))),
+    /* writes */
+    $Cons("mod1out", $Nil())
+)).
+
+// the label of "other_input" is the empty set (which makes it public).
+// so it has no label elements
+
+// Module 2
+//     reads:
+//         mod1out
+//         input4
+//     writes:
+//         mod2out 
+bindFlatParticle("module2", $FlatParticle(
+    /* reads */
+    $Cons("mod1out", $Cons("input4", $Nil())),
+    /* writes */
+    $Cons("mod2out", $Nil())
+)).
+
+//-----------------------------------------------------------------------------
+// Policy
+//-----------------------------------------------------------------------------
+
+// Assign owners to tags
+ownsTag("Alice", "raw_video").
+ownsTag("Alice", "image_selection").
+ownsTag("Alice", "product_id").
+ownsTag("ServiceProvider", "model_weights").
+
+// Assign tags to initial inputs
+claimHasTag("some_struct.a", "raw_video").
+claimHasTag("some_struct.b", "image_selection").
+claimHasTag("input4", "model_weights").
+
+// The label of mod1out is downgraded to public (= empty set) 
+// because both the raw_video and image_selection tags are removed, and then
+// upgraded to {"product_id"} with this line:
+claimHasTag("mod1out", "product_id").
+
+// Set the downgrades permitted by both parties
+saysDowngrades("Alice", "mod1out", "raw_video").
+saysDowngrades("Alice", "mod1out", "image_selection").
+saysDowngrades("ServiceProvider", "mod2out", "model_weights").
+
+//----------------------------------------------------------------------------- 
+// Policy Check
+//-----------------------------------------------------------------------------
+
+.decl test_label_check(x: number)
+
+// Define a label for doing the check of the final output
+isIFCLabel("mod2out_expected_label").
+isMember("product_id", "mod2out_expected_label").
+
+// The check passes if this is true (i.e., the CSV for test_label_check 
+// includes '1').
+// It is supposed to pass. If you comment out any of the "saysDowngrades" lines 
+// above, it should fail.
+test_label_check(1) :- accessPathLabelFlowsTo("mod2out", 
```

Comment:
Use the `testPasses` rule here? https://github.com/google-research/raksha/blob/51d56ff8f55122f6e035d99ad54a2fcc69c129a2/src/analysis/souffle/fact_test_helper.dl#L14

Reply:
OK I will try it out. I think I started this before that was done.

Reply:
done now.